eip712-adapter-moshi/src/main/java/pm/gnosis/eip712/adapters/moshi/MoshiAdapter.kt

```diff
@@ -0,0 +1,61 @@
+package pm.gnosis.eip712.adapters.moshi
+
+import com.squareup.moshi.Json
+import com.squareup.moshi.JsonClass
+import com.squareup.moshi.Moshi
+import com.squareup.moshi.Types
+import okio.buffer
+import okio.source
+import pm.gnosis.eip712.JsonAdapter
+import pm.gnosis.utils.asEthereumAddress
+import pm.gnosis.utils.nullOnThrow
+import java.io.InputStream
+
+class MoshiAdapter : JsonAdapter {
+    private val moshi by lazy { Moshi.Builder().build() }
+
+    override fun parse(rawJson: String): JsonAdapter.Result {
+        val type = Types.newParameterizedType(List::class.java, Any::class.java)
+        val root = moshi.adapter<List<Any>>(type).fromJson(rawJson) ?: throw IllegalArgumentException("Root is not a Json Array")
+        return parse(root)
+    }
+
+    override fun parse(inputStream: InputStream): JsonAdapter.Result =
+        inputStream.source().buffer().use { bufferedSource ->
+            val type = Types.newParameterizedType(List::class.java, Any::class.java)
+            val root = moshi.adapter<List<Any>>(type).fromJson(bufferedSource) ?: throw IllegalArgumentException("Root is not a Json Array")
+            return parse(root)
+        }
+
+    private fun parse(root: List<Any>): JsonAdapter.Result {
+        val signer = nullOnThrow { root[0].toString().asEthereumAddress() }
+            ?: throw IllegalArgumentException("First element is not a valid signer")
+        val typedDataAsMap = root[1] as? Map<*, *>
+            ?: throw IllegalArgumentException("Second element is not an object")
+        // Hack: converting back to json string (not the most efficient) - see https://github.com/square/moshi/issues/318
+        val rawTypedData = moshi.adapter(Map::class.java).toJson(typedDataAsMap)
+        val typedData = nullOnThrow { moshi.adapter(TypedData::class.java).fromJson(rawTypedData) }
```

Comment:
This is super ugly ......

Reply:
Yes 😞 Dealing with different types in Arrays is not pretty. I thought about a more efficient solution but also kind of ugly. I went with this one for readability.

Reply:
Fixed in https://github.com/gnosis/svalinn-kotlin/pull/31/commits/10cc16042762e85f3bc4b6a25f4241cec57d8a94